src/firefly/js/visualize/XYPlot.jsx

```diff
@@ -34,288 +35,610 @@ const plotParamsShape = PropTypes.shape({
+
+    debouncedResize() {
+        return debounce((newWidth, newHeight) => {
+            const chart = this.refs.chart && this.refs.chart.getChart();
+            if (chart) {
+                const {data} = this.props;
+                if (data.decimateKey) {
+                    // update marker's size
+                    const {xUnitPx, yUnitPx} = this.getDeciSymbolSize(chart, data.decimateKey);
+                    chart.series.forEach((series) => {
+                        series.name.includes(UNSELECTED) && series.update({
+                            marker: {radius: xUnitPx / 2, hD: (xUnitPx - yUnitPx) / 2}}, false);
+                    });
+                }
+                const {chartWidth, chartHeight} = this.calculateChartSize(newWidth, newHeight);
+
+                chart.setSize(chartWidth, chartHeight, this.shouldAnimate() );
+
+                if (data.decimate) {chart.redraw();}
             }
-            const numericData = data.reduce((numdata, arow, idx) => {
-                const nrow = arow.map(toNumber);
-                pushFunc(numdata, nrow, idx);
-                return numdata;
-            }, {selected: [], unselected: []});
+        }, 500);
+    }
+
+    shouldAnimate() {
+        const {data} = this.props;
+        return (!data || !data.rows || data.rows.length <= 250);
+    }
+
+    adjustPlotDisplay() {
+        const {params} = this.props;
+
+        // can add more chart events here like
+        // chart.bind('selection', (event) => {});
+
+        if (params.selection) {
+            this.updateSelectionRect(params.selection);
+        }
+    }
+
+    updateSelectionRect(selection) {
+        const chart = this.refs.chart.getChart();
+
+        if (this.selectionRect) {
+            this.selectionRect.destroy();
+            this.selectionRect = undefined;
+        }
+        if (selection) {
+            const xMinPx = chart.xAxis[0].toPixels(selection.xMin);
+            const xMaxPx = chart.xAxis[0].toPixels(selection.xMax);
+            const yMinPx = chart.yAxis[0].toPixels(selection.yMin);
+            const yMaxPx = chart.yAxis[0].toPixels(selection.yMax);
+            const width = Math.abs(xMaxPx - xMinPx);
+            const height = Math.abs(yMaxPx - yMinPx);
+            this.selectionRect = chart.renderer.rect(Math.min(xMinPx, xMaxPx), Math.min(yMinPx, yMaxPx), width, height, 1)
+                .attr({
+                    fill: selectionRectColor,
+                    stroke: '#8c8c8c',
+                    'stroke-width': 0.5,
+                    zIndex: 7 // same as Highcharts' selectionMrker rectangle
+                })
+                .add();
+        }
+    }
+
+    onSelectionEvent(event) {
+        const xAxis = event.xAxis[0];
+        const yAxis = event.yAxis[0];
+
+        this.props.onSelection({xMin: xAxis.min, xMax: xAxis.max, yMin: yAxis.min, yMax: yAxis.max});
+    }
+
+    makeSeries(chart) {
+        //const chart = this.refs.chart && this.refs.chart.getChart();
```

Comment:
I think this function could be broken out into a pure function and maybe even put into another file (`XYPlotHelper.js`???).  It is over 100 lines and is key to the charts.